File: python/2-D Dynamic Programming/unique_paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        """
        Find the number of unique paths from top-left to bottom-right in an m x n grid.
        You can only move down or right at any point in time.
        
        Approach:
        - Use dynamic programming with space optimization.
        - Initialize the bottom row with 1s (only one way to reach each cell in the last row).
        - For each row from bottom to top:
          - Create a new row starting with 1s.
          - For each cell (except the last), the number of paths equals:
            paths from right + paths from below (which is the value from the previous row).
        - Return the value at the top-left corner.
        
        Time Complexity: O(m * n)
        - We iterate through each cell in the grid once
        
        Space Complexity: O(n)
        - We only store two rows at a time (current and previous)
        
        Args:
            m: int - Number of rows
            n: int - Number of columns
        Returns:
            int - Number of unique paths from top-left to bottom-right
        """
        row = [1] * n  # Initialize the bottom row with 1s

        for i in range(m - 1):
            newRow = [1] * n  # Initialize new row with 1s
            for j in range(n - 2, -1, -1):
                # Number of paths = paths from right + paths from below
                newRow[j] = newRow[j + 1] + row[j]
            row = newRow  # Update the current row
        return row[0]  # Return the value at the top-left corner

# -------------------
# Test Cases
# -------------------
def test_unique_paths():
    s = Solution()
    # Test case 1: Basic case
    assert s.uniquePaths(3, 7) == 28, "Failed on (3, 7)"
    # Test case 2: Square grid
    assert s.uniquePaths(3, 3) == 6, "Failed on (3, 3)"
    # Test case 3: Single row
    assert s.uniquePaths(1, 5) == 1, "Failed on (1, 5)"
    # Test case 4: Single column
    assert s.uniquePaths(5, 1) == 1, "Failed on (5, 1)"
    # Test case 5: 2x2 grid
    assert s.uniquePaths(2, 2) == 2, "Failed on (2, 2)"
    # Test case 6: Large grid
    assert s.uniquePaths(7, 3) == 28, "Failed on (7, 3)"
    # Test case 7: Single cell
    assert s.uniquePaths(1, 1) == 1, "Failed on (1, 1)"
    print("All test cases passed!")

if __name__ == "__main__":
    test_unique_paths()
        

File: C++/2-D Dynamic Programming/unique_paths.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int uniquePaths(int m, int n) {
        /*
            Problem:
            - Count the number of unique paths from top-left to bottom-right in an m x n grid.
            - You can only move either down or right.

            Approach:
            - Use dynamic programming with space optimization.
            - We only need to keep track of the current row and the row below it.
            - Start with the bottom row initialized to 1s.
            - Work our way up the grid by filling each row based on the values from the row below and the cell to the right.

            Time Complexity: O(m * n)
                - We visit every cell exactly once.

            Space Complexity: O(n)
                - Only two rows of size n are used.
        */

        vector<int> row(n, 1);  // Initialize the bottom row with 1s (only one path to each cell in bottom row)

        // Iterate from the second-to-last row to the top row
        for (int i = 0; i < m - 1; i++) {
            vector<int> newRow(n, 1);  // Start a new row with 1s

            // Fill the row from right to left (excluding last column)
            for (int j = n - 2; j >= 0; j--) {
                // Current cell = paths from the right + paths from the cell below
                newRow[j] = newRow[j + 1] + row[j];
            }

            row = newRow;  // Move up: current row becomes the previous row
        }

        return row[0];  // Top-left corner value is the answer
    }
};

/*
================ LINE-BY-LINE SYNTAX EXPLANATION =================

#include <bits/stdc++.h>
    - Includes all standard C++ headers (vector, iostream, etc.)

using namespace std;
    - Allows you to use standard library names (like vector) without prefixing with std::.

class Solution {
    - Defines a class named Solution.

public:
    - Public access modifier. Makes the following methods accessible from outside the class.

int uniquePaths(int m, int n)
    - Function returns an integer and takes two integers (rows and columns) as input.

vector<int> row(n, 1);
    - Creates a vector of size n initialized with all elements as 1.

for (int i = 0; i < m - 1; i++)
    - Loop over each row from bottom to top (excluding bottom row already initialized).

vector<int> newRow(n, 1);
    - Creates a new vector for the current row, initialized with 1.

for (int j = n - 2; j >= 0; j--)
    - Loop over each cell in the current row from right to left, skipping the last column.

newRow[j] = newRow[j + 1] + row[j];
    - Update the current cell as the sum of right cell and the cell below.

row = newRow;
    - Assign the newly computed row to be the "previous" row for the next iteration.

return row[0];
    - Return the number of unique paths starting from the top-left corner.

*/
